src/freenas/usr/local/lib/middlewared_truenas/plugins/failover_/event_linux.py

```diff
@@ -0,0 +1,713 @@
+            logger.warning('Entering MASTER on "%s".', ifname)
+
+            # need to stop fenced just in case it's running already
+            self.run_call('failover.fenced.stop')
+
+            logger.warning('Starting fenced')
+            fenced_error = self.run_call('failover.fenced.start')
+
+        # starting fenced daemon failed....which is bad
+        # emit an error and exit
+        if fenced_error:
+            if fenced_error == 1:
+                logger.error('Failed to register keys on disks, exiting!')
+            elif fenced_error == 2:
+                logger.error('Fenced is running on the remote node, exiting!')
+            elif fenced_error == 3:
+                logger.error('10% or more of the disks failed to be reserved, exiting!')
+            elif fenced_error == 5:
+                logger.error('Fenced encountered an unexpected fatal error, exiting!')
+            else:
+                logger.error(f'Fenced exited with code "{fenced_error}" which should never happen, exiting!')
+
+            job.set_progress(None, description='ERROR')
+            raise FencedError()
+
+        # remove the zpool cache files if necessary
+        if os.path.exists(self.ZPOOL_KILLCACHE):
+            for i in (self.ZPOOL_CACHE_FILE, self.ZPOOL_CACHE_FILE_SAVED):
+                with contextlib.suppress(Exception):
+                    os.unlink(i)
+
+        # create the self.ZPOOL_KILLCACHE file
+        else:
+            with contextlib.suppress(Exception):
+                with open(self.ZPOOL_KILLCACHE, 'w') as f:
+                    f.flush()  # be sure it goes straight to disk
+                    os.fsync(f.fileno())  # be EXTRA sure it goes straight to disk
+
+        # if we're here and the zpool "saved" cache file exists we need to check
+        # if it's modify time is < the standard zpool cache file and if it is
+        # we overwrite the zpool "saved" cache file with the standard one
+        if os.path.exists(self.ZPOOL_CACHE_FILE_SAVED) and os.path.exists(self.ZPOOL_CACHE_FILE):
+            zpool_cache_mtime = os.stat(self.ZPOOL_CACHE_FILE).st_mtime
+            zpool_cache_saved_mtime = os.stat(self.ZPOOL_CACHE_FILE_SAVED).st_mtime
+            if zpool_cache_mtime > zpool_cache_saved_mtime:
+                with contextlib.suppress(Exception):
+                    shutil.copy2(self.ZPOOL_CACHE_FILE, self.ZPOOL_CACHE_FILE_SAVED)
+
+        # set the progress to IMPORTING
+        job.set_progress(None, description='IMPORTING')
+
+        failed = []
+        for vol in fobj['volumes']:
+            logger.info('Importing %s', vol['name'])
+
+            # import the zpool(s)
+            try:
+                self.run_call(
+                    'zfs.pool.import_pool',
+                    vol['guid'],
+                    {
+                        'altroot': '/mnt',
+                        'cachefile': self.ZPOOL_CACHE_FILE,
+                    }
+                )
+            except Exception as e:
```

Comment:
IIRC import_pool may raise a non-fatal error here, e.g. unable to mount some datasets

Reply:
Nope, that method catches the mount failed exceptions and logs them for us so don't need to worry about catching it here.